interfaces/prompting/patterns/patterns.go

```diff
@@ -0,0 +1,184 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package patterns
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strings"
+
+	doublestar "github.com/bmatcuk/doublestar/v4"
+)
+
+// Limit the number of expanded path patterns for a particular pattern.
+// When fully expanded, the number of patterns for a given unexpanded pattern
+// may not exceed this limit.
+const maxExpandedPatterns = 1000
+
+// PathPattern is an iterator which yields expanded path patterns.
+type PathPattern struct {
+	original   string
+	renderTree renderNode
+}
+
+// ParsePathPattern validates the given pattern and parses it into a PathPattern
+// from which expanded path patterns can be iterated, and returns it.
+func ParsePathPattern(pattern string) (*PathPattern, error) {
+	pathPattern := &PathPattern{}
+	if err := pathPattern.parse(pattern); err != nil {
+		return nil, err
+	}
+	return pathPattern, nil
+}
+
+// parse validates the given pattern and parses it into a PathPattern from
+// which expanded path patterns can be iterated, overwriting the receiver.
+func (p *PathPattern) parse(pattern string) error {
+	prefix := fmt.Sprintf("cannot parse path pattern %q", pattern)
+	if pattern == "" {
+		return fmt.Errorf("%s: pattern has length 0", prefix)
+	}
+	if pattern[0] != '/' {
+		return fmt.Errorf("%s: pattern must start with '/'", prefix)
+	}
+	if strings.HasSuffix(pattern, `\`) && !strings.HasSuffix(pattern, `\\`) {
+		return fmt.Errorf(`%s: trailing unescaped '\' character`, prefix)
+	}
```

Comment:
I think this should be caught by the scanner now?

Reply:
Yes, I agree simpler to leave it to the scanner.